planetiler-core/src/main/java/com/onthegomap/planetiler/util/Format.java

```diff
@@ -17,26 +17,40 @@
 public class Format {
 
   public static final Locale DEFAULT_LOCALE = Locale.getDefault(Locale.Category.FORMAT);
-  public static final ConcurrentMap<Locale, Format> instances = new ConcurrentHashMap<>();
-  private final NumberFormat pf;
-  private final NumberFormat nf;
-  private final NumberFormat intF;
+
+  private static final ConcurrentMap<Locale, Format> instances = new ConcurrentHashMap<>();
+
+  // `NumberFormat` instances are not thread safe, so we need to wrap them inside a `ThreadLocal`.
+  //
+  // Ignore warnings about not removing thread local values since planetiler uses dedicated worker threads that release
+  // values when a task is finished and are not re-used.
+  @SuppressWarnings("java:S5164")
+  private final ThreadLocal<NumberFormat> pf;
+  @SuppressWarnings("java:S5164")
+  private final ThreadLocal<NumberFormat> nf;
+  @SuppressWarnings("java:S5164")
+  private final ThreadLocal<NumberFormat> intF;
 
   private Format(Locale locale) {
-    pf = NumberFormat.getPercentInstance(locale);
-    pf.setMaximumFractionDigits(0);
-    nf = NumberFormat.getNumberInstance(locale);
-    nf.setMaximumFractionDigits(1);
-    intF = NumberFormat.getNumberInstance(locale);
-    intF.setMaximumFractionDigits(0);
+    pf = ThreadLocal.withInitial(() -> {
+      var f = NumberFormat.getPercentInstance(locale);
+      f.setMaximumFractionDigits(0);
+      return f;
+    });
+    nf = ThreadLocal.withInitial(() -> {
+      var f = NumberFormat.getNumberInstance(locale);
+      f.setMaximumFractionDigits(1);
+      return f;
+    });
+    intF = ThreadLocal.withInitial(() -> {
+      var f = NumberFormat.getNumberInstance(locale);
+      f.setMaximumFractionDigits(0);
+      return f;
+    });
```

Comment:
minor thing - you could probably just inline these in the field definitions since they don't need anything outside the initializer

Reply:
Is that possible if they need access to the `locale` parameter passed to the constructor?

Reply:
oh yeah, good point!